### ISSUE TITLE
Lack of SVG Fill Color Customization

### ISSUE THREAD

**Issue by itsAdee** [typst/typst#1939]
### Description

Greetings,
I'm reaching out regarding an issue I've come across while using Typst for document writing. Currently, there seems to be a limitation in changing the fill (color) of an SVG image within Typst. I'm keen to inquire if there's an existing method to modify the fill color, or if not, whether there's potential for implementing this feature. I am also interested in taking on the task of addressing this issue and would greatly appreciate any guidance on the best approach to achieve this enhancement.

### Use Case

Users often require icons to complement text in their presentations. By enabling SVG fill color changes, icons can be more deliberately aligned with text aesthetics, enhancing the overall visual impact of documents.

**Comment by LaurenzV** [comment]
Is there an SVG-specific feature that allows changing all colors used inside of the SVG? As far as I'm aware, there isn't so the only way to achieve this would be to replace all "fill" attributes in the SVG with the color you want to fill it with, which I think would be a very brittle approach. So I'm not sure how this should be implemented on our side. 

Do you have an example of an SVG file you want to recolor?

**Comment by itsAdee** [comment]
I was writing a document on material 3 for a university assessment and wanted to use some material 3 icons from the material 3 website : https://fonts.google.com/icons, I have also attached a material 3 clock svg file as an example.
![clock](https://github.com/typst/typst/assets/104891437/5937f8d4-000a-489d-abaa-3127959d7ce9)

**Comment by LaurenzV** [comment]
I see. Well, currently this is not possible without resorting to hacks. I think from the technical perspective, it would be possible to implement, but I think the main issue is in the details of how this should be implemented in practice:

1. How should the user be able to indicate the color? Should there be another attribute in the `image` function? If so, this would mean that the function has an additional attribute which only works for SVG files and not raster images, which is not so nice in my opinion.
2. Which colours should exactly be replaced? Should only simple fill colours be replaced or also fills like gradients and patterns? What about fills inside of masks?

But I think this feature has been requested before, so maybe it does mean that many people want it and we should add it. But for now, in the case of simple SVGs like the one you posted, it should suffice if you just change it manually by specifying the color yourself. In your case, changing it to the following will make the whole thing blue. You can change the color by changing it inside of the "fill" attribute:

```svg
<svg xmlns="http://www.w3.org/2000/svg" height="48" viewBox="0 -960 960 960" width="48">
<path fill="blue" d="M480-80q-82 0-155-31.5t-127.5-86Q143-252 111.5-325T80-480q0-83 31.5-156t86-127Q252-817 325-848.5T480-880q83 0 156 31.5T763-763q54 54 85.5 127T880-480q0 29-4 58t-12 56q-11-11-24.5-19T810-397q5-20 7.5-40.5T820-480q0-142-99-241t-241-99q-141 0-240.5 99T140-480q0 141 99.5 240.5T480-140q58 0 110-18.5t95-50.5q10 12 22 22t26 16q-52 43-116.5 67T480-80Zm300-161q-17 0-28.5-11.5T740-281q0-17 11.5-28.5T780-321q17 0 28.5 11.5T820-281q0 17-11.5 28.5T780-241Zm-153-46L453-468v-225h60v201l159 160-45 45Z"/>
</svg>
```

**Comment by itsAdee** [comment]
Alright. Thank you

**Comment by Beiri22** [comment]
Does this solve your issue?

**Comment by itsAdee** [comment]
No, the decode method does not seem to work for the vg, I download from the material 3 website, I tried changing the black color to another color but it remains the same ,however , I downloaded a random svg from the internet and the methods works correctly on it,  the code i used to text the random svg from the internet : #let original = read("clock2.svg")
#let changed = original.replace(
  "#000000",
  green.hex(),
)

#image.decode(original,width:25pt)
#image.decode(changed,width:25pt)
"
![image](https://github.com/typst/typst/assets/104891437/8eb62d0b-8da2-4481-8cf0-56ea005df15c)
The code i used to text the clock svg from the material 3 website : "#let original = read("clock.svg")
#let changed = original.replace(
  "#000000",
  green.hex(),
)

#image.decode(original,width:25pt)
#image.decode(changed,width:25pt)"
![image](https://github.com/typst/typst/assets/104891437/28e62f13-3576-4eab-80dd-412fb13bb19d)
But the colour does not change here , can this somehow be fixed, attached both the icons as well
![clock2](https://github.com/typst/typst/assets/104891437/5a03057d-8d09-434c-95b5-b157bc8a69e6)
![clock](https://github.com/typst/typst/assets/104891437/284fcbae-20aa-4e43-89ff-cd06b74494df)

**Comment by Beiri22** [comment]
The non working file does not contain any color information. Black seems to be the default, not explicitly mentioned. I'll have a look...

**Comment by itsAdee** [comment]
Thank you

**Comment by Beiri22** [comment]
```typ
#let colorize(svg, color) = {
  let blk = black.hex();
  // You might improve this prototypical detection.
  if svg.contains(blk) { 
    // Just replace
    svg.replace(blk, color.hex())
  } else {
    // Explicitly state color
    svg.replace("<svg ", "<svg fill=\""+color.hex()+"\" ")
  }
}


= Old
#let c1 = read("c1.svg");
#let c2 = read("c2.svg");
#let c1 = c1.replace("#000000",green.hex())
#let c2 = c2.replace("#000000",green.hex())
#image.decode(c1,width:25pt)
#image.decode(c2,width:25pt)

= New
#let c1 = read("c1.svg");
#let c2 = read("c2.svg");
#let c1 = colorize(c1, green);
#let c2 = colorize(c2, green);
#image.decode(c1,width:25pt)
#image.decode(c2,width:25pt)
```
 try that ... it's prototypical though

**Comment by itsAdee** [comment]
will do

**Comment by itsAdee** [comment]
it works thanks again

**Comment by NicoWeio** [comment]
Note: `color.hex()` is now `color.to-hex()`.

**Comment by LaurenzV** [comment]
For your information, I'm working on a plugin that will allow you to do that easily. I hope that I will be able to get it done within the next few weeks/months.

**Comment by LaurenzV** [comment]
The idea is to make a plugin that allows to preprocess SVGs. Currently, this will probably only include two things, but this list might be extended if I get new ideas:
- Allow changing the color of all strokes/fills.
- Allow changing the text font/size/etc. of all text elements.

It will be a real WASM plugin and use [usvg](https://github.com/RazrFalcon/resvg) to read the SVG into memory, change the properties accordingly and then write it again. The advantage of doing it this way is that it is much more reliable (e.g. it doesn't matter whether the current fill color is specified as `fill="color"` or via some CSS propery, as long as it's a valid SVG, it will handle it properly).

**Comment by Beiri22** [comment]
i understand, it seems suitable. Will you post a link here, once ready?

**Comment by LaurenzV** [comment]
Sure. 👍

**Comment by laurmaedje** [comment]
The only way I could see this being supported natively is with SVG's `currentColor`, but I think it's unsupported in usvg and it's also unclear what the API would be in Typst. The string-level preprocessing or plugin approach is quite flexible, so should we just close this?

**Comment by LaurenzV** [comment]
The best way would be using [stylesheet injection](https://github.com/RazrFalcon/resvg/pull/785) and then creating a Typst plugin that uses that to provide the functionality. Maybe at some point I will keep working on it.

But I agree that this is out of scope for Typst itself, so we can close this.

**Comment by JeppeKlitgaard** [comment]
> The idea is to make a plugin that allows to preprocess SVGs. Currently, this will probably only include two things, but this list might be extended if I get new ideas:
> 
> * Allow changing the color of all strokes/fills.
> * Allow changing the text font/size/etc. of all text elements.
> 
> It will be a real WASM plugin and use [usvg](https://github.com/RazrFalcon/resvg) to read the SVG into memory, change the properties accordingly and then write it again. The advantage of doing it this way is that it is much more reliable (e.g. it doesn't matter whether the current fill color is specified as `fill="color"` or via some CSS propery, as long as it's a valid SVG, it will handle it properly).



@LaurenzV, This sounds like a really good idea. I would love to have a little bit more control over images directly in Typst – did you ever get around to implementing this?

**Comment by LaurenzV** [comment]
I haven't unfortunately. :(

**Comment by Pyrrhu5** [comment]
Since `image.decode` is being depreciated, the snippet has to be update to:

```typst
let image_raw = read(image_path)
image_raw = image_raw.replace("currentColor", red.to-hex())
image_raw = bytes(image_raw)
let icon_image = image(image_raw, width: 1em)
```

**Comment by ptrn23** [comment]
thanks! just noticed `image.decode` is being depreciated too while going down this thread, really helped a lot!

**Comment by krakowski** [comment]
@Pyrrhu5 Thanks! This is really helpful. I created a small reusable function from your example for anyone interested.

```typ
#let colorized-image(path, color: black, ..args) = {
  let raw = read(path).replace("currentColor", color.to-hex())
  image(bytes(raw), ..args.named())
}
```

**Comment by teohhanhui** [comment]
Actually that snippet doesn't / can't do the right thing:

> In the [color](https://drafts.csswg.org/css-color/#propdef-color) property, the used value of currentcolor is the resolved [inherited value](https://drafts.csswg.org/css-cascade-5/#inherited-value). In any other property, its used value is the used value of the color property on the same element.
>
> Note: This means that if the [currentcolor](https://drafts.csswg.org/css-color/#valdef-color-currentcolor) value is inherited, it’s inherited as a keyword, not as the value of the [color](https://drafts.csswg.org/css-color/#propdef-color) property, so descendants will use their own color property to resolve it.

https://drafts.csswg.org/css-color/#resolving-other-colors
